File: src/main/java/org/nanowrimo/ws/RegionWordCountHandler.java
package org.nanowrimo.ws;

import org.xml.sax.Attributes;
import org.xml.sax.SAXException;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

@SuppressWarnings({"unused", "Duplicates"})
class RegionWordCountHandler extends WordCountHandler<RegionWordCount> {
    private StringBuffer sb = new StringBuffer();
    private RegionWordCount regionSummary;
    private Date entryDate;
    private int wordcount;
    private boolean insideEntry;
    private SimpleDateFormat sdf;
    private List<WordCountListEntry> entries;

    public RegionWordCountHandler() {
        sdf = new SimpleDateFormat("yyyy-MM-dd");
        entries = new ArrayList<>();
        regionSummary = new RegionWordCount();
    }

    @Override
    RegionWordCount getSummary() {
        return regionSummary;
    }

    public RegionWordCount getRegionSummary() {
        return regionSummary;
    }

    public void characters(char ch[], int start, int length) {
        sb.append(ch, start, length);
    }

    public void startElement(String uri, String localName, String qName, Attributes attributes) {
        sb.setLength(0);
        if (WORD_COUNT_ENTRY.equals(qName)) {
            insideEntry = true;
        }
    }

    public void endElement(String uri, String localName, String qName) {
        if (REGION_NAME.equals(qName)) {
            regionSummary.setName(sb.toString());
        } else if (WORD_COUNT_ENTRY.equals(qName)) {
            insideEntry = false;
            if (entryDate != null) {
                WordCountListEntry entry = new WordCountListEntry();
                entry.setDate(entryDate);
                entry.setWordcount(wordcount);
                entries.add(entry);
            }
        } else if (WORD_COUNT_ENTRY_DATE.equals(qName)) {
            try {
                entryDate = sdf.parse(sb.toString());
            } catch (ParseException e) {
                e.printStackTrace();
            }
        } else if (REGION_ID.equals(qName)) {
            try {
                regionSummary.setRegionID(Integer.parseInt(sb.toString()));
            } catch (NumberFormatException ignored) {
            }
        } else if (WORD_COUNT_GOAL.equals(qName)) {
            try {
                regionSummary.setGoal(Integer.parseInt(sb.toString()));
            } catch (NumberFormatException ignored) {
            }
        } else if (REGION_AVERAGE.equals(qName)) {
            try {
                regionSummary.setAverage(Integer.parseInt(sb.toString()));
            } catch (NumberFormatException ignored) {
            }
        } else if (REGION_COUNT.equals(qName)) {
            try {
                regionSummary.setCount(Integer.parseInt(sb.toString()));
            } catch (NumberFormatException ignored) {
            }
        } else if (REGION_STDDEV.equals(qName)) {
            try {
                regionSummary.setStddev(Integer.parseInt(sb.toString()));
            } catch (NumberFormatException ignored) {
            }
        } else if (REGION_MIN.equals(qName)) {
            try {
                regionSummary.setMin(Integer.parseInt(sb.toString()));
            } catch (NumberFormatException ignored) {
            }
        } else if (REGION_MAX.equals(qName)) {
            try {
                regionSummary.setMax(Integer.parseInt(sb.toString()));
            } catch (NumberFormatException ignored) {
            }
        } else if (WORD_COUNT_RANK.equals(qName)) {
            try {
                regionSummary.setRank(Integer.parseInt(sb.toString()));
            } catch (NumberFormatException ignored) {
            }
        } else if (WORD_COUNT.equals(qName) && insideEntry) {
            try {
                wordcount = Integer.parseInt(sb.toString());
            } catch (NumberFormatException ignored) {
            }
        } else if (REGION_WORD_COUNT.equals(qName)) {
            try {
                regionSummary.setWordcount(Integer.parseInt(sb.toString()));
            } catch (NumberFormatException ignored) {
            }
        } else if (WORDCOUNTS_CONTAINER.equals(qName)) {
            regionSummary.setEntries(entries.toArray(new WordCountListEntry[entries.size()]));
        }
    }
}

File: src/main/java/org/nanowrimo/ws/RegionWordCountHandler.kt
package org.nanowrimo.ws

import org.xml.sax.Attributes

import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.ArrayList
import java.util.Date

internal class RegionWordCountHandler : WordCountHandler<RegionWordCount>() {
    private val sdf: SimpleDateFormat = SimpleDateFormat("yyyy-MM-dd")
    private val sb = StringBuffer()
    private var entryDate: Date? = null
    private var wordcount: Int = 0
    private var insideEntry: Boolean = false
    private val entries: MutableList<WordCountListEntry>
    override val summary: RegionWordCount

    init {
        entries = ArrayList()
        summary = RegionWordCount()
    }

    override fun characters(ch: CharArray?, start: Int, length: Int) {
        sb.append(ch, start, length)
    }

    override fun startElement(uri: String?, localName: String?, qName: String?, attributes: Attributes?) {
        sb.setLength(0)
        if (WORD_COUNT_ENTRY == qName) {
            insideEntry = true
        }
    }

    override fun endElement(uri: String?, localName: String?, qName: String?) {
        when {
            REGION_NAME == qName -> summary.name = sb.toString()
            WORD_COUNT_ENTRY == qName -> {
                insideEntry = false
                if (entryDate != null) {
                    val entry = WordCountListEntry()
                    entry.date = entryDate
                    entry.wordcount = wordcount
                    entries.add(entry)
                }
            }
            WORD_COUNT_ENTRY_DATE == qName -> try {
                entryDate = sdf.parse(sb.toString())
            } catch (e: ParseException) {
                e.printStackTrace()
            }
            REGION_ID == qName -> try {
                summary.regionID = sb.toString().toInt()
            } catch (ignored: NumberFormatException) {
            }
            WORD_COUNT_GOAL == qName -> try {
                summary.goal = sb.toString().toInt()
            } catch (ignored: NumberFormatException) {
            }
            REGION_AVERAGE == qName -> try {
                summary.average = sb.toString().toInt()
            } catch (ignored: NumberFormatException) {
            }
            REGION_COUNT == qName -> try {
                summary.count = sb.toString().toInt()
            } catch (ignored: NumberFormatException) {
            }
            REGION_STDDEV == qName -> try {
                summary.stddev = sb.toString().toInt()
            } catch (ignored: NumberFormatException) {
            }
            REGION_MIN == qName -> try {
                summary.min = sb.toString().toInt()
            } catch (ignored: NumberFormatException) {
            }
            REGION_MAX == qName -> try {
                summary.max = sb.toString().toInt()
            } catch (ignored: NumberFormatException) {
            }
            WORD_COUNT_RANK == qName -> try {
                summary.rank = sb.toString().toInt()
            } catch (ignored: NumberFormatException) {
            }
            WORD_COUNT == qName && insideEntry -> try {
                wordcount = sb.toString().toInt()
            } catch (ignored: NumberFormatException) {
            }
            REGION_WORD_COUNT == qName -> try {
                summary.wordcount = sb.toString().toInt()
            } catch (ignored: NumberFormatException) {
            }
            WORDCOUNTS_CONTAINER == qName -> summary.entries = entries
        }
    }
}
